src/operator/numpy/np_tensordot_op-inl.h

```diff
@@ -0,0 +1,575 @@
+    } else if (!aT && bT) {
+      CHECK_EQ(ad2, bd2);
+      Tensor<xpu, 2, DType> out_tensor = out.get_with_shape<xpu, 2, DType>(Shape2(ad1, bd1), s);
+      ASSIGN_DISPATCH(out_tensor, req, dot(a_tensor, b_tensor.T()));
+    } else {
+      CHECK_EQ(ad2, bd1);
+      Tensor<xpu, 2, DType> out_tensor = out.get_with_shape<xpu, 2, DType>(Shape2(ad1, bd2), s);
+      ASSIGN_DISPATCH(out_tensor, req, dot(a_tensor, b_tensor));
+    }
+  });
+}
+
+/**
+ * Calculates tensordot.
+ */
+template<typename xpu>
+void TensordotImpl(
+    const Tuple<int>& a_axes_summed,
+    const Tuple<int>& b_axes_summed,
+    const OpContext& ctx,
+    const TBlob& a,
+    const TBlob& b,
+    const TBlob& out,
+    const std::vector<OpReqType>& req) {
+
+  if (req[0] == kNullOp) {
+    return;
+  }
+
+  if (out.shape_.Size() == 0U) {
+    return;  // zero-size output, no need to launch kernel
+  }
+
+  const mxnet::TShape& a_shape = a.shape_;
+  const mxnet::TShape& b_shape = b.shape_;
+
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  CHECK_EQ(out.type_flag_, a.type_flag_)
+      << "Binary function only support input/output with the same type";
+  CHECK_EQ(out.type_flag_, b.type_flag_)
+      << "Binary function only support input/output with the same type";
+  CHECK(out.type_flag_ == kFloat32 || out.type_flag_ == kFloat64 ||
+      (out.type_flag_ == kFloat16 && ctx.run_ctx.ctx.dev_mask() == mshadow::gpu::kDevMask))
+      << "Tensordot only supports float32/float64 for CPU, and float16/float32/float64 for GPU";
+
+  Tuple<int> a_axes_remained;
+  Tuple<int> b_axes_remained;
+  Tuple<int> a_axes;
+  Tuple<int> b_axes;
+  GetReorderedAxes(a_axes_summed, &a_axes_remained, &a_axes, b_axes_summed, &b_axes_remained,
+                   &b_axes, a_shape, b_shape);
+
+  int ad1 = 1, ad2 = 1, bd1 = 1, bd2 = 1;
+  GetMatrixDimensions(&ad1, &ad2, &bd1, &bd2, a_axes_remained, a_axes_summed,
+                      b_axes_remained, b_axes_summed, a_shape, b_shape);
+
+  mxnet::TShape a_temp_shape = GetReorderedShape(a_shape, a_axes);
+  mxnet::TShape b_temp_shape = GetReorderedShape(b_shape, b_axes);
+
+  MSHADOW_REAL_TYPE_SWITCH(out.type_flag_, DType, {
+    if (a_shape.Size() == 0U || b_shape.Size() == 0U) {  // 0-size input
+      if (req[0] != kAddTo) {
+        Tensor<xpu, 1, DType> out_data = out.get_with_shape<xpu, 1, DType>(
+            Shape1(out.shape_.Size()), s);
+        out_data = static_cast<DType>(0);
+      }
+      return;
+    }
+
+    Tensor<xpu, 1, DType> workspace = ctx.requested[0].get_space_typed<xpu, 1, DType>
+      (Shape1(a.Size() + b.Size()), s);
+    DType* a_ptr = reinterpret_cast<DType*>(workspace.dptr_);
+    DType* b_ptr = reinterpret_cast<DType*>(workspace.dptr_ + a.Size());
+    TBlob a_res = TBlob(a_ptr, a_temp_shape, xpu::kDevMask);
+    TBlob b_res = TBlob(b_ptr, b_temp_shape, xpu::kDevMask);
+
+    mxnet::op::TransposeImpl<xpu>(ctx.run_ctx, a, a_res,
+        mxnet::TShape(a_axes.begin(), a_axes.end()));
```

Comment:
```c++
    mxnet::op::TransposeImpl<xpu>(ctx.run_ctx, a, a_res,
                                  mxnet::TShape(a_axes.begin(), a_axes.end()));
```
Similar comment for function call below.

Reply:
Fixed.Thx.